crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -228,22 +217,57 @@ impl DunderAllKind<'_> {
     }
 }
 
-/// Given an array of strings, return `true` if they are already
-/// ordered according to an isort-style sort.
-fn dunder_all_is_already_sorted(string_elements: &[&str]) -> bool {
-    let mut element_iter = string_elements.iter();
-    let Some(this) = element_iter.next() else {
-        return true;
-    };
-    let mut this_key = AllItemSortKey::from(*this);
-    for next in element_iter {
-        let next_key = AllItemSortKey::from(*next);
-        if next_key < this_key {
-            return false;
+/// Determine whether an `__all__` tuple/list is sorted,
+/// unsorted, or invalid. If it's unsorted, determine whether
+/// there's a possibility that we could generate a fix for it.
+///
+/// ("Sorted" here means "ordered according to an isort-style sort".
+/// See the module-level docs for a definition of "isort-style sort.")
+#[derive(Debug, is_macro::Is)]
+enum DunderAllSortClassification<'a> {
+    Sorted,
+    Invalid,
```

Comment:
Nit: Maybe `Unsupported`. It's a valid `__all__` expression, but not one the fix supports

Reply:
Hmm, maybe `Unsortable` or `NotAStringLiteralList` would be best. We return this variant for truly invalid things like

```py
__all__ = [1, 2, 3]
```

 But we also return this variant for things that might or might not be valid, like

```py
foo = "whatever"
__all__ = [foo]
```

The thing that these have in common is that they're beyond the scope of this rule (not just the fix), so, while they might or might not be valid `__all__` definitions, we don't emit a violation for them

Reply:
I'll go with `Unsortable` 👍